File: reference-api/src/main/js/model/ParseContext.js
import _ from 'underscore';
import ParseCursor from "./ParseCursor";
import {Constant, Variable, Brackets,Ternary, createExpression, getReturnType} from "./ExpressionModel";
import {NodeTypes, ReturnTypes,HighlightTypes} from '../Constants';
import {validateCompletenessOfArguments} from './ParseUtil';
import SuggestionModelType from './SuggestionModelType';
import SuggestionModel from "./SuggestionModel";


const LOG = false;
const LONG_PATTERN = /^(-?[0-9]+)/;
const DOUBLE_PATTERN = /^(-?[0-9]+\.[0-9]+)/;
const TEXT_PATTERN  = /^'((\S|\s)+?)'/;
const VARIABLE_PATTERN  = /^\$([0-9A-Za-z\\.]+)/;
const BRACKET_OPEN = "(";
const PARAMETER_SEPARATOR = ",";
const BRACKET_CLOSE = ")";

function log(msg){
    if(LOG) console.log(msg);
}


export function ParseContext(text){
    const cursor= new ParseCursor(text);
    const parseStackStash = this.parseStackStash = [];
    let parseStack = [];
    const blocks = this.blocks = [];
    this.suggestions = new SuggestionModel();
    this.cursor= cursor;
    this.unaryLib={};
    this.infixLib={};
    this.functionLib={};

    function push(item){
        parseStack.push(item);
    }

    function popStack() {
        if(parseStack.length!==0){
            throw "Bug";
        }
        parseStack=parseStackStash.pop();
    }

    function stashStack() {
        parseStackStash.push(parseStack);
        parseStack=[];

    }

    function addBlock(block, type){
        blocks.push({start: block.start, end: block.end, type: type});
    }


    function justSawConstant() {
        if(parseStack.length === 0) return false;
        return _.last(parseStack).type === NodeTypes.CONSTANT;
    }

    this.done=function() {
        return !this.cursor.more();
    };

    this.libScan=function(lib, tryParse, addSuggestions){
        const operators = _.sortBy(_.keys(lib), e => -e.length);
        const any = _.any(operators, e=>{
            if(tryParse.call(this, e)) return true;
        });
        this.suggestions.add(addSuggestions.call(this));
        return any;
    };

    this.tryVariable=function() {
        if (cursor.at(VARIABLE_PATTERN)) {
            const txt = cursor.read(VARIABLE_PATTERN);
            push(new Variable(txt.substring(1)),cursor.getLastBlock());
            addBlock(cursor.getLastBlock(),HighlightTypes.VARIABLE);
            return true;
        }
        return false;
    };

    this.tryBooleanConstant=function() {
        if (cursor.at("true")){
            cursor.read("true");
            push(new Constant(true, ReturnTypes.BOOLEAN));
            addBlock(cursor.getLastBlock(),HighlightTypes.CONSTANT);
            return true;
        }
        if (cursor.at("false")){
            cursor.read("false");
            push(new Constant(false, ReturnTypes.BOOLEAN));
            addBlock(cursor.getLastBlock(),HighlightTypes.CONSTANT);
            return true;
        }
        return false;
    };

    this.tryLongConstant = function() {
        if (!justSawConstant() &&
            !cursor.at(DOUBLE_PATTERN) &&
            cursor.at(LONG_PATTERN)) {
            const txt = cursor.read(LONG_PATTERN);
            push(new Constant(parseInt(txt), ReturnTypes.INTEGER));
            addBlock(cursor.getLastBlock(),HighlightTypes.CONSTANT);
            return true;
        }
        return false;
    };

    this.tryDoubleConstant=function() {
        if (!justSawConstant() &&
            cursor.at(DOUBLE_PATTERN)) {
            const txt = cursor.read(DOUBLE_PATTERN);
            push(new Constant(parseFloat(txt), ReturnTypes.DECIMAL));
            addBlock(cursor.getLastBlock(),HighlightTypes.CONSTANT);
            return true;
        }
        return false;
    };

    this.tryStringConstant=function() {
        if (cursor.at(TEXT_PATTERN)) {
            const txt = cursor.read(TEXT_PATTERN);
            push(new Constant(txt, ReturnTypes.TEXT));
            addBlock(cursor.getLastBlock(),HighlightTypes.CONSTANT);
            return true;
        }
        return false;
    };

    this.tryFunction=function(functionName) {
        if (cursor.at(functionName)) {
            cursor.read(functionName);
            addBlock(cursor.getLastBlock(),HighlightTypes.FUNCTION);
            cursor.read(BRACKET_OPEN);
            addBlock(cursor.getLastBlock(),HighlightTypes.FUNCTION_SYMBOLS);
            const expr = createExpression(this.functionLib[functionName]);
            const argumentsTypes = expr.argumentsTypes;
            stashStack();
            for (let i = 0; i< argumentsTypes.length; i++){
                const separators = [BRACKET_CLOSE, PARAMETER_SEPARATOR];
                this.parseExpression(separators);
                expr.arguments.push(this.yield(expr.priority()));
                if(cursor.at(BRACKET_CLOSE)){
                    validateCompletenessOfArguments(expr);
                    cursor.read(BRACKET_CLOSE);
                    addBlock(cursor.getLastBlock(),HighlightTypes.FUNCTION_SYMBOLS);
                    break;
                }
                cursor.read(PARAMETER_SEPARATOR);
                addBlock(cursor.getLastBlock(),HighlightTypes.FUNCTION_SYMBOLS);
            }
            popStack();
            push(expr);
            return true;
        }
        return false;
    };

    this.tryInfix=function(operator){
        if (cursor.at(operator) && !this.empty()) {
            cursor.read(operator);
            const expr = createExpression(this.infixLib[operator]);
            addBlock(cursor.getLastBlock(),HighlightTypes.INFIX);
            this.reduce(expr.priority());
            const lhs = parseStack.pop();
            log("POP "+lhs.type);
            expr.arguments.push(lhs);
            push(expr);
            return true;
        }
        return false;
    };

    this.tryUnary=function(operator){
        if (cursor.at(operator)) {
            cursor.read(operator);
            addBlock(cursor.getLastBlock(),HighlightTypes.INFIX);
            const expr = createExpression(this.unaryLib[operator]);
            push(expr);
            return true;
        }
        return false;
    };

    this.tryBrackets=function() {
        if (cursor.at(BRACKET_OPEN)) {
            cursor.read(BRACKET_OPEN);
            addBlock(cursor.getLastBlock(),HighlightTypes.BRACKETS);
            this.parseExpression(BRACKET_CLOSE);
            const inner = this.yield();
            const brackets = Brackets();
            brackets.arguments.push(inner);
            push(brackets, "()");
            cursor.read(BRACKET_CLOSE);
            addBlock(cursor.getLastBlock(),HighlightTypes.BRACKETS);
            return true;
        }
        return false;
    };

     this.tryTernary=function() {
        if (cursor.at("?") && !this.empty()) {
            cursor.read("?");
            addBlock(cursor.getLastBlock(),HighlightTypes.TERNARY);
            const ternary = Ternary();
            const condition = this.yield(ternary.priority());
            ternary.arguments.push(condition);
            push(ternary);
            this.parseExpression(":");
            cursor.read(":");
            addBlock(cursor.getLastBlock(),HighlightTypes.TERNARY);
            return true;
        }
        return false;
    };

    this.reduce=function(targetPrio)
    {
        if (parseStack.length >= 2)
        {
            const argument = parseStack.pop();
            const operator = _.last(parseStack);
            let argumentPriority = argument.priority();
            let operatorPriority = operator.priority();
            let reduce = operatorPriority>=targetPrio && operator.argumentsTypes!==undefined;

            if(argument.type === NodeTypes.TERNARY &&
                operator.type === NodeTypes.TERNARY &&
                targetPrio === 0
            ){
                reduce = false;
            }

            if(LOG) {
                log(">>>>>>>>>");
                log("NEW: prio " + targetPrio+"");
                log("ARG:" + argument.operator + "(prio " + argumentPriority+")");
                log("OPR:" + operator.operator + "(prio " + operatorPriority+")");
                if(reduce) log("REDUCE");
                log("<<<<<<<<");

            }
            if(reduce){
                operator.arguments.push(argument);
                this.reduce(targetPrio);
            } else {
                parseStack.push(argument);
                //consume argument
            }
        }
    };


    this.withUnaryLib=function(unaryLib){
        this.unaryLib=unaryLib;
        return this;
    };
    this.withFunctionLib=function(functionLib){
        this.functionLib=functionLib;
        return this;
    };

    this.withInfixLib=function(infixLib){
        this.infixLib=infixLib;
        return this;
    };

    this.parseExpression=function(until){
        while(!this.done() && (until===null ||  until===undefined || !this.cursor.at(until))) {
            log("========");
            this.suggestions.clear()
            if( this.tryLongConstant() ||
                this.tryBooleanConstant() ||
                this.tryDoubleConstant() ||
                this.tryStringConstant() ||
                this.tryVariable() ||
                this.tryBrackets() ||
                this.tryTernary() ||
                this.libScan(this.unaryLib, this.tryUnary, this.unarySuggestions) ||
                this.libScan(this.functionLib, this.tryFunction, this.functionSuggestions) ||
                this.libScan(this.infixLib, this.tryInfix, this.infixSuggestions)){
                this.suggestions.clear();
                continue;
            }
            throw "Can't match head"
        }
        this.reduce(until==null?-1:0);
    };
    this.getBlocks=function(){
        return this.blocks;
    };
    this.getSuggestions=function(){
        return this.suggestions;
    };
    this.yield=function(prio){
        this.reduce(prio||0);
        return parseStack.pop();
    };

    this.empty=function(){
        return parseStack.length===0;
    };

    this.functionSuggestions=function(){
        var names = _.sortBy(_.keys(this.functionLib), i=>i);
        const suggestions =   _.map(names, name=>{ return {name: name, type: getReturnType(this.functionLib[name].returnType)}});
        return new SuggestionModelType('functions', suggestions);
    };

    this.infixSuggestions=function(){
        var names = _.sortBy(_.keys(this.infixLib), i=>i);
        const suggestions =   _.map(names, name=>{ return {name: name, type: getReturnType(this.infixLib[name].returnType)}});
        return new SuggestionModelType('infix', suggestions);
    };

    this.unarySuggestions=function(){
        const names = _.sortBy(_.keys(this.unaryLib), i=>i);
        const suggestions =  _.map(names, name=>{ return {name: name, type: getReturnType(this.unaryLib[name].returnType)}});
        return new SuggestionModelType('unary', suggestions);
    };

    this.constantSuggestions=function(){
        return new SuggestionModelType('constants', null, [
            {name: 'integer', type:  ReturnTypes.INTEGER, pattern: LONG_PATTERN},
            {name: 'decimal', type: ReturnTypes.DECIMAL, pattern: DOUBLE_PATTERN},
            {name: 'text', type: ReturnTypes.TEXT, pattern: TEXT_PATTERN},
            {name: 'boolean', type: ReturnTypes.BOOLEAN, pattern: /^(true|false)/}
        ]);
    };
}



export default ParseContext

File: json-templates/src/main/java/nl/softcause/jsontemplates/expresionparser/ParseContext.java
package nl.softcause.jsontemplates.expresionparser;

import static nl.softcause.jsontemplates.expresionparser.ParseUtils.createExpression;
import static nl.softcause.jsontemplates.expresionparser.ParseUtils.operator;

import java.util.Collections;
import java.util.Comparator;
import java.util.Map;
import java.util.Stack;
import java.util.function.Function;
import java.util.regex.Pattern;

import lombok.NonNull;
import nl.softcause.jsontemplates.expressions.*;
import nl.softcause.jsontemplates.expressions.logic.Ternary;
import nl.softcause.jsontemplates.types.IExpressionType;

class ParseContext {

    private static final boolean LOG = false;
    private static final Pattern LONG_PATTERN = Pattern.compile("^(-?[0-9]+)");
    private static final Pattern DOUBLE_PATTERN = Pattern.compile("^(-?[0-9]+\\.[0-9]+)");
    //    private static final Pattern TEXT_PATTERN = Pattern.compile("^'((\\w|\\s|[-!?.\";:{}/\\\\#])+)'");
    private static final Pattern TEXT_PATTERN = Pattern.compile("^'((?:[^'\\\\]|\\\\.)*)'");
    //    /'(?:[^'\\]|\\.)*'/
    private static final Pattern VARIABLE_PATTERN = Pattern.compile("^\\$([0-9A-Za-z\\.]+)");
    private static final String BRACKET_OPEN = "(";
    private static final String PARAMETER_SEPARATOR = ",";
    private static final String BRACKET_CLOSE = ")";
    private static final Pattern NULL_TRAIL = Pattern.compile("^null([0-9A-Za-z\\.]+)");
    private static final Pattern TRUE_TRAIL = Pattern.compile("^true([0-9A-Za-z\\.]+)");
    private static final Pattern FALSE_TRAIL = Pattern.compile("^false([0-9A-Za-z\\.]+)");

    private Stack<IExpression> parseStack = new Stack<>();
    private final Stack<Stack<IExpression>> parseStackStash = new Stack<>();

    private final ParseCursor cursor;
    private Map<String, Class> unaryLib;
    private Map<String, Class> functionLib;
    private Map<String, Class> infixLib;

    ParseContext(@NonNull String text) {
        cursor = new ParseCursor(text);
    }

    ParseContext withFunctionLib(Map<String, Class> lib) {
        this.functionLib = lib;
        return this;
    }

    ParseContext withInfixLib(Map<String, Class> lib) {
        this.infixLib = lib;
        return this;
    }

    public ParseContext withUnaryLib(Map<String, Class> lib) {
        this.unaryLib = lib;
        return this;
    }

    void parseExpression(String[] until) {
        while (!done() && (until == null || !cursor.at(until))) {
            log("========");
            if (tryNullConstant() ||
                    tryLongConstant() ||
                    tryBooleanConstant() ||
                    tryDoubleConstant() ||
                    tryStringConstant() ||
                    tryVariable() ||
                    tryBrackets() ||
                    tryTernary() ||
                    libScan(unaryLib, this::tryUnary) ||
                    libScan(functionLib, this::tryFunction) ||
                    libScan(infixLib, this::tryInfix)) {
                continue;
            }
            throw ParseException.cantMatchHead().at(cursor);
        }
        reduce(until == null ? -1 : 0);
    }

    private boolean libScan(Map<String, Class> lib, Function<String, Boolean> tryParse) {
        var operators = lib
                .keySet()
                .stream()
                .sorted(Comparator.comparingInt(e -> -e.length()))
                .toArray(String[]::new);

        for (var e : operators) {
            if (tryParse.apply(e)) {
                return true;
            }
        }
        return false;
    }

    private boolean tryVariable() {
        if (cursor.at(VARIABLE_PATTERN)) {
            var txt = cursor.read(VARIABLE_PATTERN);
            push(new Variable(txt));
            return true;
        }
        return false;
    }

    private boolean tryNullConstant() {
        if (cursor.at("null") && !cursor.at(NULL_TRAIL)) {
            cursor.read("null");
            push(new Constant(null));
            return true;
        }
        return false;
    }

    private boolean tryBooleanConstant() {
        if (cursor.at("true") && !cursor.at(TRUE_TRAIL)) {
            cursor.read("true");
            push(new Constant(true));
            return true;
        }
        if (cursor.at("false") && !cursor.at(FALSE_TRAIL)) {
            cursor.read("false");
            push(new Constant(false));
            return true;
        }
        return false;
    }

    private boolean tryLongConstant() {
        if (!justSawConstant() &&
                !cursor.at(DOUBLE_PATTERN) &&
                cursor.at(LONG_PATTERN)) {
            var txt = cursor.read(LONG_PATTERN);
            push(new Constant(Long.parseLong(txt)));
            return true;
        }
        return false;
    }

    private boolean justSawConstant() {
        if (parseStack.empty()) {
            return false;
        }
        return parseStack.peek() instanceof Constant;
    }


    private boolean tryDoubleConstant() {
        if (!justSawConstant() &&
                cursor.at(DOUBLE_PATTERN)) {
            var txt = cursor.read(DOUBLE_PATTERN);
            push(new Constant(Double.parseDouble(txt)));
            return true;
        }
        return false;
    }


    private boolean tryStringConstant() {
        if (cursor.at(TEXT_PATTERN)) {
            var txt = cursor.read(TEXT_PATTERN);
            push(new Constant(txt));
            return true;
        }
        return false;
    }


    private boolean tryFunction(String functionName) {
        if (cursor.at(functionName)) {
            cursor.read(functionName);
            cursor.read(BRACKET_OPEN);
            var expr = (IExpressionWithArguments) createExpression(functionLib.get(functionName));
            IExpressionType[] argumentsTypes = expr.getArgumentsTypes();
            stashStack();
            for (var i = 0; i < argumentsTypes.length; i++) {
                var separators = new String[] {BRACKET_CLOSE, PARAMETER_SEPARATOR};
                parseExpression(separators);
                expr.getArguments().add(yield(expr.priority()));
                if (cursor.at(BRACKET_CLOSE)) {
                    ParseUtils.validateCompletenessOfArguments(expr, cursor);
                    cursor.read(BRACKET_CLOSE);
                    break;
                }
                cursor.read(PARAMETER_SEPARATOR);
            }
            popStack();
            push(expr);
            return true;
        }
        return false;
    }

    private void popStack() {
        if (!parseStack.empty()) {
            throw new RuntimeException("Bug");
        }
        parseStack = parseStackStash.pop();
    }

    private void stashStack() {
        parseStackStash.push(parseStack);
        parseStack = new Stack<>();

    }
//
//    private static void validateCompletenessOfArgumetns(IExpressionWithArguments expr, IExpressionType[] argumentsTypes, ParseCursor cursor) {
//        for (var j = expr.getArguments().size(); j < argumentsTypes.length; j++){
//            if(!(argumentsTypes[j] instanceof nl.softcause.jsontemplates.types.Optional)){
//                throw ParseException.expectedMoreArguments().at(cursor);
//            }
//        }
//    }

    private boolean tryInfix(String operator) {
        if (cursor.at(operator) && !empty()) {
            cursor.read(operator);
            var expr = (IExpressionWithArguments) createExpression(infixLib.get(operator));
            reduce(expr.priority());
            var lhs = parseStack.pop();
            log("POP " + operator(lhs));
            expr.getArguments().add(lhs);
            push(expr);
            return true;
        }
        return false;
    }

    private boolean tryUnary(String operator) {
        if (cursor.at(operator)) {
            cursor.read(operator);
            var expr = (IExpressionWithArguments) createExpression(unaryLib.get(operator));
            push(expr);
            return true;
        }
        return false;
    }

    private boolean tryBrackets() {
        if (cursor.at(BRACKET_OPEN)) {
            cursor.read(BRACKET_OPEN);
            parseExpression(new String[] {BRACKET_CLOSE});
            var inner = yield();
            var brackets = new Brackets();
            brackets.setArguments(Collections.singletonList(
                    inner
            ));
            push(brackets);
            cursor.read(BRACKET_CLOSE);
            return true;
        }
        return false;
    }


    private boolean tryTernary() {
        if (cursor.at("?") && !empty()) {
            cursor.read("?");
            var ternary = new Ternary();
            var condition = yield(ternary.priority());
            ternary.getArguments().add(condition);
            push(ternary);
            parseExpression(new String[] {":"});
            cursor.read(":");
            return true;
        }
        return false;
    }

    IExpression yield() {
        reduce(0);
        return parseStack.pop();
    }

    IExpression yield(int targetPriority) {
        reduce(targetPriority);
        return parseStack.pop();
    }

    private boolean done() {
        return !cursor.more();
    }

    boolean empty() {
        return parseStack.isEmpty();
    }

    private void push(IExpression expression) {
        log("PUSH " + operator(expression) + " " + cursor);
        parseStack.push(expression);
    }

    private void reduce(int targetPrio) {
        if (parseStack.size() >= 2) {
            IExpression argument = parseStack.pop();
            IExpression operator = parseStack.peek();
            var argumentPriority = argument.priority();
            var operatorPriority = operator.priority();
            var reduce = operatorPriority >= targetPrio && operator instanceof IExpressionWithArguments;

            if (argument instanceof Ternary &&
                    operator instanceof Ternary &&
                    targetPrio == 0
            ) {
                reduce = false;
            }

            if (LOG) {
                System.out.println(">>>>>>>>>");
                System.out.println("NEW: prio " + targetPrio + "");
                System.out.println("ARG:" + operator(argument) + "(prio " + argumentPriority + ")");
                System.out.println("OPR:" + operator(operator) + "(prio " + operatorPriority + ")");
                if (reduce) {
                    System.out.println("REDUCE");
                }
                System.out.println("<<<<<<<<");

            }
            if (reduce) {
                ((IExpressionWithArguments) operator).getArguments().add(argument);
                reduce(targetPrio);
            } else {
                parseStack.push(argument);
                //consume argument
            }
        }
    }

    private void log(String msg) {
        if (LOG) {
            System.out.println(msg);
        }
    }

    protected ParseCursor getCursor() {
        return cursor.clone();
    }

}
